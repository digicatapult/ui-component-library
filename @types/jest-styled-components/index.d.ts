declare module 'jest-styled-components/serializer'
