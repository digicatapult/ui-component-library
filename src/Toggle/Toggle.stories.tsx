import React, { useState } from 'react'
import { Story } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import Toggle from './index.js'

export default {
  title: 'Components/Toggle',
  component: Toggle,
  argTypes: {
    height: {
      control: 'text',
    },
    width: {
      control: 'text',
    },
    borderRadius: {
      control: 'text',
    },
    padding: {
      control: 'number',
    },
    offBackground: {
      control: 'color',
    },
    onBackground: {
      control: 'color',
    },
    buttonColor: {
      control: 'color',
    },
  },
}

const DefaultStoryTemplate: Story = (args) => {
  return <Toggle onChange={action('change')} {...args} />
}

export const Default = DefaultStoryTemplate
export const Square = DefaultStoryTemplate.bind({})
Square.args = {
  borderRadius: '0',
}

export const ControlledInput: Story = (args) => {
  const [isChecked, setIsChecked] = useState(true)

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setIsChecked(e.target.checked)
  }

  return <Toggle onChange={onChange} checked={isChecked} {...args} />
}
