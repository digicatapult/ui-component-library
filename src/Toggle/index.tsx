import React from 'react'
import styled from 'styled-components'

export interface ToggleProps {
  width?: string
  height?: string
  borderRadius?: string
  padding?: number
  offBackground?: string
  onBackground?: string
  buttonColor?: string
  checked?: boolean
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void
}

const Toggle: React.FC<ToggleProps> = ({
  width = '100px',
  height = '50px',
  borderRadius = '20px',
  padding = 5,
  offBackground = 'lightgray',
  onBackground = 'green',
  buttonColor = 'white',
  checked,
  onChange,
}) => (
  <>
    <HiddenCheckbox
      type="checkbox"
      id="switch"
      checked={checked}
      onBackground={onBackground}
      padding={padding}
      onChange={(e) => onChange?.(e)}
    />
    <Label
      width={width}
      height={height}
      borderRadius={borderRadius}
      padding={padding}
      offBackground={offBackground}
      onBackground={onBackground}
      buttonColor={buttonColor}
      htmlFor="switch"
    >
      Toggle
    </Label>
  </>
)

const HiddenCheckbox = styled.input<ToggleProps>`
  height: 0;
  width: 0;
  visibility: hidden;

  :checked + label {
    background: ${({ onBackground }) => onBackground};
  }

  :checked + label:after {
    left: calc(100% - ${({ padding }) => padding}px);
    transform: translateX(-100%);
  }
`

const Label = styled.label<ToggleProps>`
  cursor: pointer;
  text-indent: -9999px;
  width: ${({ width }) => width};
  height: ${({ height }) => height};
  background: ${({ offBackground }) => offBackground};
  display: block;
  border-radius: ${({ borderRadius }) => borderRadius};
  position: relative;

  :after {
    content: '';
    position: absolute;
    top: ${({ padding }) => padding}px;
    left: ${({ padding }) => padding}px;
    width: 40%;
    height: calc(100% - ${({ padding }) => (padding || 0) * 2}px);
    background: ${({ buttonColor }) => buttonColor};
    border-radius: ${({ borderRadius }) => borderRadius};
    transition: 0.3s;
  }

  :active:after {
    width: 60%;
  }
`

export default Toggle
