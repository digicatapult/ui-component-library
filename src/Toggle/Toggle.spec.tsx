import React from 'react'
import renderer, { act } from 'react-test-renderer'
import 'jest-styled-components'

import Toggle from './index.js'

describe('Test', () => {
  test('default', () => {
    const changeHandler = jest.fn(() => {})
    const tree = renderer.create(<Toggle onChange={changeHandler} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with props', () => {
    const changeHandler = jest.fn(() => {})
    const tree = renderer
      .create(
        <Toggle
          height="2.5em"
          width="2.5em"
          offBackground="#27847A"
          onBackground="#000"
          borderRadius="1.25em"
          onChange={changeHandler}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('click', async () => {
    const changeHandler = jest.fn((e: React.ChangeEvent) => {
      return e
    })
    const Wrapper = () => {
      return <Toggle onChange={changeHandler} />
    }

    const instance = renderer.create(<Wrapper />)

    await act(() => {
      instance.root.findByType('input').props.onChange('test')
    })

    expect(changeHandler.mock.calls).toHaveLength(1)

    const change = changeHandler.mock.calls[0]
    expect(change).toEqual(['test'])
  })
})
