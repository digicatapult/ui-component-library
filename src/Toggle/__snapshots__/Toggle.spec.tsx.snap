// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test default 1`] = `
[
  .c0 {
  height: 0;
  width: 0;
  visibility: hidden;
}

.c0:checked + label {
  background: green;
}

.c0:checked + label:after {
  left: calc(100% - 5px);
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

<input
    className="c0"
    id="switch"
    onChange={[Function]}
    type="checkbox"
  />,
  .c0 {
  cursor: pointer;
  width: 100px;
  height: 50px;
  background: lightgray;
  display: block;
  border-radius: 20px;
  position: relative;
}

.c0:after {
  content: '';
  position: absolute;
  top: 5px;
  left: 5px;
  width: 40%;
  height: calc(100% - 10px);
  background: white;
  border-radius: 20px;
  -webkit-transition: 0.3s;
  transition: 0.3s;
}

.c0:active:after {
  width: 60%;
}

<label
    className="c0"
    height="50px"
    htmlFor="switch"
    width="100px"
  />,
]
`;

exports[`Test with props 1`] = `
[
  .c0 {
  height: 0;
  width: 0;
  visibility: hidden;
}

.c0:checked + label {
  background: #000;
}

.c0:checked + label:after {
  left: calc(100% - 5px);
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
}

<input
    className="c0"
    id="switch"
    onChange={[Function]}
    type="checkbox"
  />,
  .c0 {
  cursor: pointer;
  width: 2.5em;
  height: 2.5em;
  background: #27847A;
  display: block;
  border-radius: 1.25em;
  position: relative;
}

.c0:after {
  content: '';
  position: absolute;
  top: 5px;
  left: 5px;
  width: 40%;
  height: calc(100% - 10px);
  background: white;
  border-radius: 1.25em;
  -webkit-transition: 0.3s;
  transition: 0.3s;
}

.c0:active:after {
  width: 60%;
}

<label
    className="c0"
    height="2.5em"
    htmlFor="switch"
    width="2.5em"
  />,
]
`;
